src/bin/cat.rs

```diff
@@ -3,30 +3,207 @@
 extern crate extra;
 
 use std::env;
+use std::error::Error;
 use std::fs;
-use std::io;
+use std::io::{self, Read, Stderr, StdoutLock, Write};
 use extra::option::OptionalExt;
 
-fn main() {
-    let paths = env::args().skip(1);
-    let stdin = io::stdin();
-    let mut stdin = stdin.lock();
-    let stdout = io::stdout();
-    let mut stdout = stdout.lock();
-    let mut stderr = io::stderr();
+const MAN_PAGE: &'static str = r#"NAME
+    cat - concatenate files and print on the standard output
+
+SYNOPSIS
+    cat [-h | --help] [-A | --show-all] [-b | --number-nonblank] [-e] [-E | --show-ends] [-n | --number] [-s | --squeeze-blank] [-t] [-T] FILES...
+
+DESCRIPTION
+    Concatenates all files to the standard output.
+
+    If no file is given, or if FILE is '-', read from standard input.
+
+OPTIONS
+    -A
+    --show-all
+        equivalent to -v -E -T
+
+    -b
+    --number-nonblank
+        number nonempty output lines, overriding -n
+
+    -e
+        equivalent to -v -E
+
+    -E
+    --show-ends
+        display $ at the end of each line
+
+    -n
+    --number
+        number all output lines
+
+    -s
+    --squeeze-blank
+        supress repeated empty output lines
+
+    -t
+        equivalent to -v -T
+
+    -T
+    --show_tabs
+        display TAB characters as ^I
 
-    if paths.len() == 0 {
-        io::copy(&mut stdin, &mut stdout).try(&mut stderr);
-    } else {
-        for path in paths {
-            if path == "-" {
-                io::copy(&mut stdin, &mut stdout).try(&mut stderr);
+    -v
+    --show-nonprinting
+        use ^ and M- notation,except for LFD and TAB.
+
+    -h
+    --help
+        display this help and exit
+
+AUTHOR
+    Written by Michael Murphy.
+"#;
+
+struct Program {
+    number:          bool,
+    number_nonblank: bool,
+    show_ends:       bool,
+    show_tabs:       bool,
+    squeeze_blank:   bool,
+    paths:           Vec<String>,
+}
+
+impl Program {
+    /// Initialize the program's arguments and flags.
+    fn initialize(stdout: &mut StdoutLock, stderr: &mut Stderr) -> Program {
+        let mut cat = Program {
+            number:          false,
+            number_nonblank: false,
+            show_ends:       false,
+            show_tabs:       false,
+            squeeze_blank:   false,
+            paths:           Vec::new()
+        };
+        for arg in env::args().skip(1) {
+            if arg.starts_with('-') {
+                match arg.as_str() {
+                    "-h" | "--help" => {
+                        stdout.write(MAN_PAGE.as_bytes()).try(stderr);
+                        stdout.flush().try(stderr);
+                        std::process::exit(0);
+                    }
+                    "-A" | "--show-all" => {
+                        // TODO: implement -v
```

Comment:
Doesn't seem to print `^I` which is a tab. I see no difference in the output of cat -v versus cat.